### PR TITLE
BoxOperations: Consider amount for change box

### PR DESCRIPTION
**Pull request by MrStahlfelge**
When we have tokens in selected inboxes that shouldn't be sent, a change box is needed to return the tokens to the sending address. That means that a small amount of ERG is needed on top of the actual amount to send which must be considered when selecting boxes.

This change adds the ability to BoxOperations to account for a change box amount and an accompanying test case. However, the test case files although BoxOperations is now correctly selecting the two boxes of the test case. This is due to `DefaultBoxSelector` ignoring the second, needed box which needs to be fixed in ergo-wallet.

Follow-up to #103

### CODE REVIEW

File: common/src/main/java/org/ergoplatform/appkit/SelectTokensHelper.java [hunk 1]
```diff
@@ -10,33 +10,60 @@
...
         }
         return tokensNeeded;
     }
 
+    /**
+     * Marks the given tokens as needed, subtracting the amount from the remaining amount of
+     * tokens to find.
+     * Also updates if a change box is needed
+     */
+    public void useTokens(Iterable<ErgoToken> foundTokens) {
```
Comment (aslesarenko) [inline common/src/main/java/org/ergoplatform/appkit/SelectTokensHelper.java:51]: Please update the doc. It is not clear how tokens are marked. What is means "tokens to find"? Find where?

File: common/src/main/java/org/ergoplatform/appkit/SelectTokensHelper.java [hunk 2]
```diff
@@ -58,4 +85,8 @@ public boolean areTokensCovered() {
         }
         return result;
     }
+
+    public boolean isChangeBoxNeeded() {
```
Comment (aslesarenko) [inline common/src/main/java/org/ergoplatform/appkit/SelectTokensHelper.java:89]: Please add JavaDoc.

File: common/src/main/java/org/ergoplatform/appkit/SelectTokensHelper.java [hunk 3]
```diff
@@ -10,33 +10,60 @@
...
+                if (newValue < 0) {
+                    changeBoxNeeded = true;
+                }
+            } else {
+                changeBoxNeeded = true;
+            }
+        }
+    }
+
     public boolean areTokensCovered() {
```
Comment (aslesarenko) [inline common/src/main/java/org/ergoplatform/appkit/SelectTokensHelper.java:67]: Please add JavaDoc

File: common/src/main/java/org/ergoplatform/appkit/SelectTokensHelper.java [hunk 4]
```diff
@@ -10,33 +10,60 @@
  */
 public class SelectTokensHelper {
```
Comment (aslesarenko) [inline common/src/main/java/org/ergoplatform/appkit/SelectTokensHelper.java:11]: Please clarify the description. What "tokens already found" means?

File: lib-api/src/main/java/org/ergoplatform/appkit/BoxOperations.java [hunk 5]
```diff
@@ -157,18 +159,27 @@ public String send(
...
 
         for (Address sender : senders) {
             inputBoxesLoader.prepareForAddress(sender);
-            CoveringBoxes unspent = getCoveringBoxesFor(remainingAmount, remainingTokens,
+            CoveringBoxes addressUnspentBoxes = getCoveringBoxesFor(remainingAmount, remainingTokens,
+                changeBoxConsidered,
                 page -> inputBoxesLoader.loadBoxesPage(ctx, sender, page));
-            for (InputBox b : unspent.getBoxes()) {
+
+            // change box needed but not considered yet?
```
Comment (aslesarenko) [inline lib-api/src/main/java/org/ergoplatform/appkit/BoxOperations.java:174]: What it means "considered"? Maybe different verb will serve better.

File: lib-api/src/main/java/org/ergoplatform/appkit/BoxOperations.java [hunk 6]
```diff
@@ -255,13 +266,15 @@ public static SignedTransaction spendBoxesTx(
      * - returning an empty list means the source of input boxes is drained and no further page will
      *   be loaded
      *
-     * @param amountToSpend    amount of NanoErgs to be covered
-     * @param tokensToSpend    ErgoToken to spent
-     * @param inputBoxesLoader method returning paged sets of InputBoxes, see above
+     * @param amountToSpend       amount of NanoErgs to be covered
+     * @param tokensToSpend       ErgoToken to spent
+     * @param changeBoxConsidered true if change box is already considered by amountToSpend
```
Comment (aslesarenko) [inline lib-api/src/main/java/org/ergoplatform/appkit/BoxOperations.java:271]: the word `considered` is not clear.

**Review by aslesarenko — CHANGES_REQUESTED** [review]

**Comment by MrStahlfelge** [comment]
Addressed the feedback, please check

**Review by aslesarenko — APPROVED** [review]

**Pull request merged**